core/src/xmake/machine.c

```diff
@@ -260,75 +291,73 @@ static tb_size_t xm_machine_get_program_file(xm_machine_impl_t* impl, tb_char_t*
-        if (!_NSGetExecutablePath(path, &bufsize))
-            ok = tb_true;
-#elif defined(TB_CONFIG_OS_LINUX)
-        // get the executale file path as program directory
-        ssize_t size = readlink("/proc/self/exe", path, (size_t)maxn);
-        if (size > 0 && size < maxn)
+        tb_uint32_t size = (tb_uint32_t)maxn;
+        if (!_NSGetExecutablePath(data, &size))
         {
-            // end
-            path[size] = '\0';
+            // trace
+            tb_trace_d("programfile: %s", data);
 
-            // ok
-            ok = tb_true;
-        }
-#endif
-
-    } while (0);
+            // set _EXECUTABLE_PATH
+            lua_pushstring(impl -> lua, data);
+            lua_setglobal(impl -> lua, "_EXECUTABLE_PATH");
 
-    // ok?
-    if (ok)
-    {
-        // trace
-        tb_trace_d("programfile: %s", path);
+            if (!ok)
+            {
+                // get path size
+                size = tb_strlen(data);
 
-        // save the directory to the global variable: _PROGRAM_FILE
-        lua_pushstring(impl->lua, path);
-        lua_setglobal(impl->lua, "_PROGRAM_FILE");
-    }
+                // get the directory
+                while (size-- > 0)
+                {
+                    if (data[size] == '/')
+                    {
+                        data[size] = '\0';
+                        break;
+                    }
+                }
 
-    // ok?
-    return ok;
-}
-static tb_bool_t xm_machine_get_program_directory(xm_machine_impl_t* impl, tb_char_t* path, tb_size_t maxn, tb_char_t const* programfile)
-{
-    // check
-    tb_assert_and_check_return_val(impl && path && maxn, tb_false);
+                // copy to path
+                tb_strcpy(path, data);
 
-    // done
-    tb_bool_t ok = tb_false;
-    do
-    {
-        // get it from the environment variable first
-        tb_char_t data[TB_PATH_MAXN] = {0};
-        if (tb_environment_first("XMAKE_PROGRAM_DIR", data, sizeof(data)) && tb_path_absolute(data, path, maxn))
-        {
-            // ok
-            ok = tb_true;
-            break;
+                // ok
+                ok = tb_true;
+            }
         }
-
-        // get it from program file path
-        if (programfile)
+#elif defined(TB_CONFIG_OS_LINUX)
+        // get the executale file path as program directory
+        ssize_t size = readlink("/proc/self/exe", data, (size_t)maxn);
+        if (size > 0 && size < maxn)
         {
-            tb_size_t size = tb_strlcpy(data, programfile, sizeof(data));
-            if (size < sizeof(data))
+            // end
+            data[size] = '\0';
+
+            // trace
+            tb_trace_d("programfile: %s", data);
+
+            // set _EXECUTABLE_PATH
+            lua_pushstring(impl -> lua, data);
+            lua_setglobal(impl -> lua, "_EXECUTABLE_PATH");
+
+            if (!ok)
             {
                 // get the directory
                 while (size-- > 0)
                 {
```

Comment:
I think you can use this [implementation in dev](https://github.com/tboox/xmake/blob/dev/core/src/xmake/machine.c) to replace it directly. 😸